LULU filter by C.H. Rohwer -- nonlinear filter which seems to work nicely on the d-term

pjpei opened the pull request:
Hi!

As in the title :)

It's an implementation of the LULU filter by C.H. Rohwer at a university in our country.
For reference: https://en.wikipedia.org/wiki/Lulu_smoothing

I've optimized it to O(N) complexity and it seems to work OK now.  It's still quite slow so **check CPU use before flying**.

I've enabled it on the d-term only.  If anyone wants a gyro filter with this, I checked it and I don't think it's the best use of it, but perhaps my experience was not the most representative so let me know?

In the CLI or on the OSD you can set a new lowpass filter type, LULU.  In the advanced PID options on the OSD, right at the end there's now a LULU N value.  Set N to an appropriate value, see the CPU usage, test, repeat.  It's a global N value as a two LULU filters over each other with different N values just keep the highest N value, it's the nature of the mathematics.  Don't do two lulu filters, it's supposed to do nothing except add delay (assuming I implemented correctly).

I've added OSD elements and a debug trace in case someone wants to see what the D-term looks like before and after the d-term.  Debug on the pitch and roll only, before and after d-term filter.  The debug bit still needs to be properly tested, it was a late night.

![Screenshot from 2022-01-30 06-46-36](https://user-images.githubusercontent.com/25793978/151687816-d5008687-25ba-4452-a7ad-807dc97b4ab0.png)
Above is too little LULU.  It makes squarewave type stuff on the d-term. N=6 .

![Screenshot from 2022-01-30 06-47-40](https://user-images.githubusercontent.com/25793978/151687828-e9d9a648-c8bc-4c22-a7ec-077f8cce6bbc.png)
This is a better amount of LULU, 8kpid N=12.  Still not perfect, but given that this is the only static filter I use on the above in addition to dynamic filter and kalman, think that's quite alright.

If someone wants to test it, here it is.  I believe the delay in the filter is approximately N+1 samples, so if you're running 8kpid with lulu_n_val at 12, that's 13/8000 = 1.625ms delay for this filter, which I believe is quite alright!

I've found my MATEKF411 target likes N=6 or N=7 at 4k pid, while my HGLRCF722 is alright with N=12 about at 8kpid.  Graphs from HGLRCF722, don't have a logger on the other one.

Kind regards,
pjpei

Commit:
Added OSD elements
Added debug info
Removed LULU on gyro

Commit:
Fixed up #ifdef for LULU filters

skylupop commented on src/main/common/filter.h:
```diff
@@ -55,12 +55,32 @@ typedef struct ptnFilter_s {
     FILTER_BIQUAD,
     FILTER_PT2,
     FILTER_PT3,
-    FILTER_PT4,
```
the trailing comma is used just for these situations, to make an addition to not cause a change in an already existing line.
You should leave it there (and put one on your entry too)

pjpei replied:
I'll fix this up, thank you and appreciated!

skylupop commented on src/main/common/filter.c:
```diff
@@ -273,3 +273,103 @@ FAST_CODE float ptnFilterApply(ptnFilter_t *filter, float input) {
 	  return filter->state[filter->order];
 } // ptnFilterApply
+
+void luluFilterInit(luluFilter_t *filter, int N)
```
I would create a new file dedicated to this filter.

pjpei replied:
Will do!

skylupop commented on src/main/cms/cms_menu_imu.c:
```diff
@@ -101,7 +104,10 @@ static const char * const cms_mixerImplTypeLabels[] = {
 
 static const char * const cms_FilterType[] = {
-    "PT1", "BIQUAD", "PT2", "PT3", "PT4",
+    "PT1", "BIQUAD", "PT2", "PT3", "PT4"
```
You should leave the trailing comma, is there for a reason (explained below) :)

pjpei replied:
Apologies, will fix

skylupop reviewed (commented):
Just a couple of idea/considerations. I'm going to try this filter on a mine feature where I need a good dTerm filtering!

Commit:
Split off LULU into another file
Fixed IMU menu
Fixed trailing commas

pjpei commented:
I added the changes to my repo as requested, and fixed an OSD issue I caused that didn't allow LULU to be selected as a DTERM filter from the OSD.  It was only accessible via CLI.

Commit:
Added LULU filters

pjpei commented:
> Just a couple of idea/considerations. I'm going to try this filter on a mine feature where I need a good dTerm filtering!

Let me know how it goes! I'm unreasonably excited about this, it took me a month to optimize until it worked fast enough for use in a flight controller at reasonable speed.

pjpei commented on src/main/interface/msp.c:
```diff
@@ -1313,6 +1313,9 @@ bool mspProcessOutCommand(uint8_t cmdMSP, sbuf_t *dst) {
         sbufWriteU8(dst, currentPidProfile->axis_lock_multiplier);
         sbufWriteU8(dst, currentPidProfile->emuGravityGain);
+#ifdef USE_LULU
+        sbufWriteU8(dst, currentPidProfile->lulu_n_val);
```
Sorry, spent too little time looking at this.  Do you have other features for a new version of MSP that also needs to be added?  Let me know, as I suspect we'll need to get them all in at once.

pjpei replied:
I've been asked for more comments in the lulu code too, will do both over the weekend.  Sorry for this, didn't know this would be a problem.

pjpei replied:
Removed the MSP stuff for now.

Commit:
Removed MSP changes (sorry)

Commit:
Replaced the lulu algorithm with the full Roadmaker's algorithm
to reduce aliasing and remove all bump sizes up to N
Heavily optimized Roadmaker's algorithm

Commit:
Further optimized lulu operation
Added lulu.c to speed-optimized part of compiled code

Commit:
Updated min/max functionality

nerdCopter commented on src/main/common/filter.c:
```diff
@@ -273,3 +273,5 @@ FAST_CODE float ptnFilterApply(ptnFilter_t *filter, float input) {
 
 	  return filter->state[filter->order];
 } // ptnFilterApply
+
+
```
```suggestion
```

nerdCopter commented on src/main/common/filter.h:
```diff
@@ -61,6 +61,9 @@ typedef enum {
     FILTER_PT3,
     FILTER_PT4,
+#ifdef USE_LULU   
+	FILTER_LULU,
```
```suggestion
    FILTER_LULU,
```

nerdCopter commented on src/main/interface/settings.c:
```diff
@@ -282,7 +282,10 @@ static const char * const lookupTableRcInterpolationChannels[] = {
 };
 static const char * const lookupTableFilterType[] = {
-    "PT1", "BIQUAD", "PT2", "PT3", "PT4",
+    "PT1", "BIQUAD", "PT2", "PT3", "PT4"
+#ifdef USE_LULU
+    , "LULU"
+#endif
```
```suggestion
    "PT1", "BIQUAD", "PT2", "PT3", "PT4",
#ifdef USE_LULU
    "LULU",
#endif
```

nerdCopter replied:
```suggestion
    "PT1", "BIQUAD", "PT2", "PT3", "PT4",
#ifdef USE_LULU
    "LULU",
#endif
```

nerdCopter commented on src/main/flight/pid.c:
```diff
@@ -274,6 +281,10 @@ void pidInitFilters(const pidProfile_t *pidProfile) {
                 ptnFilterInit(&dtermLowpass[axis].ptnFilter, FILTER_PT2, pidProfile->dFilter[axis].dLpf, dT);
                 break;
+            case FILTER_LULU:
+            	dtermLowpassApplyFn = (filterApplyFnPtr)luluFilterApply;
+            	luluFilterInit(&dtermLowpass[axis].luluFilter, pidProfile->lulu_n_val);
+            	break;
```
```suggestion
                dtermLowpassApplyFn = (filterApplyFnPtr)luluFilterApply;
                luluFilterInit(&dtermLowpass[axis].luluFilter, pidProfile->lulu_n_val);
                break;
```

nerdCopter commented on src/main/fc/fc_core.c:
```diff
@@ -33,6 +33,10 @@
+#ifdef USE_LULU
+#include "common/lulu.h"
+#endif
+
```
```suggestion
```

nerdCopter commented on src/main/common/lulu.h:
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+
+
```
```suggestion
```

nerdCopter commented on src/main/common/lulu.h:
```diff
@@ -0,0 +1,16 @@
+    float luluInterimB[32];
+} luluFilter_t;
+
+
```
```suggestion
```

nerdCopter commented:
all tabs in `src/main/common/lulu.c` should be convereted to 4-spaces.

Quick-Flash commented:
This LULU needs to be made up to date with the LULU in Inav as its more up to date.

pjpei committed:
Updated LULU to latest version

pjpei committed:
Updated document spacing

pjpei committed:
Fixed indentation and spacing

pjpei committed:
PROTOTYPE CODE USE AT OWN RISK
Added the LULU filter as an option for D-term and gyro filter
It's garbage for gyro filtering.  Don't do it.
Set the DTerm filter to type LULU in the CLI.  Not accessible via GUI.
This has been fixed to N=10 in filter.h, changeable to anything you like
PROTOTYPE CODE USE AT OWN RISK

pjpei commented:
I've rebased, still to test.  Please don't merge yet!

nerdCopter commented on src/main/flight/pid.c:
```diff
@@ -299,6 +310,10 @@ void pidInitFilters(const pidProfile_t *pidProfile) {
                 ptnFilterInit(&dtermLowpass2[axis].ptnFilter, FILTER_PT2, pidProfile->dFilter[axis].dLpf2, dT);
                 break;
+            case FILTER_LULU:
+            	dtermLowpass2ApplyFn = (filterApplyFnPtr)luluFilterApply;
+            	luluFilterInit(&dtermLowpass2[axis].luluFilter, pidProfile->lulu_n_val);
+            	break;
```
```suggestion
                dtermLowpass2ApplyFn = (filterApplyFnPtr)luluFilterApply;
                luluFilterInit(&dtermLowpass2[axis].luluFilter, pidProfile->lulu_n_val);
                break;
```

nerdCopter reviewed (changes requested):
i've batched together a few tab-to-spaces recommendations.

pjpei committed:
Update src/main/flight/pid.c

Co-authored-by: nerdCopter <56646290+nerdCopter@users.noreply.github.com>

pjpei committed:
Fixed spaces

pjpei commented:
Thanks for the review, appreciate it!

nerdCopter committed:
Update src/main/common/lulu.h

nerdCopter committed:
Update src/main/common/lulu.h

nerdCopter committed:
Update src/main/common/filter.c

nerdCopter committed:
Update src/main/fc/fc_core.c

pjpei commented on src/main/sensors/gyro_filter_impl.h:
```diff
@@ -36,10 +36,8 @@ static FAST_CODE void GYRO_FILTER_FUNCTION_NAME(gyroSensor_t *gyroSensor) {
-        gyroADCf = gyroSensor->lowpass2FilterApplyFn((filter_t *)&gyroSensor->lowpass2Filter[axis], gyroADCf);
-#endif
         gyroADCf = gyroSensor->lowpassFilterApplyFn((filter_t *)&gyroSensor->lowpassFilter[axis], gyroADCf);
+        gyroADCf = gyroSensor->lowpass2FilterApplyFn((filter_t *)&gyroSensor->lowpass2Filter[axis], gyroADCf);
```
Apologies, will fix

pjpei committed:
Fixed gyro filter impl to original version

Quick-Flash commented:
Please change the code from Allman indentation to K&R indentation style. https://en.wikipedia.org/wiki/Indentation_style

pjpei committed:
Fixed minor bug in code
Updated indentation

pjpei committed:
Improved on algorithm

pjpei commented:
Tested this morning, pretty happy with it.  If you're happy, I'm happy to get it merged. :)

Apologies if I missed any tabs-to-spaces things, please note if I did, will search again.

pjpei commented:
It actually works best on gyro, and a simple pt1 lowpass on dterm sorts out all the relevant noise on there on all my quads.  Might differ for others, but it's quite alright.  Note it runs slowly so check CPU use when testing!!

nerdCopter commented on src/main/io/osd.c:
```diff
@@ -219,7 +219,10 @@ static const uint8_t osdElementDisplayOrder[] = {
-    OSD_COMPASS_BAR
+    OSD_COMPASS_BAR,
+#ifdef USE_LULU
+    OSD_LULU_N
```
```suggestion
    OSD_LULU_N,
```

nerdCopter commented:
> It actually works best on gyro, [...]

you mentioned gyro, but it seems this code is only for dTerm, correct?

nerdCopter commented:
@Quick-Flash , many things are gated `#ifdef USE_LULU` here, but should "everything" be gated? for example debug, struct, source.mk and partial OSD are not.

pjpei committed:
Update src/main/io/osd.c

Co-authored-by: nerdCopter <56646290+nerdCopter@users.noreply.github.com>

pjpei commented:
> > It actually works best on gyro, [...]
> 
> you mentioned gyro, but it seems this code is only for dTerm, correct?

Hi!

I did lots of tests, but the bottom line is one LULU on gyro, one dynamic filter one notch and one PT1 on dterm sorts out the filtering on all my crafts tested.  40mm, 2", 3", 5".   I did experimental tests on removing the gyrofiltering before calculating the slope, then doing LULU on the "raw" derivative, which worked very well too.  This was not implemented, so I think only gyro is best option here.

nerdCopter commented:
- fair enough, so will you add gyro capabilities here?
- full-disclosure, there are no plans to release any further EmuFlight . i just thought this was interesting after noticing it was tested on @iNavFlight
- it would require Configurator support as well, but it is currently impossible to build configurator as the OSX builds are broken.

pjpei commented:
> * fair enough, so will you add gyro capabilities here?
Yeah, I think it's already available to put on the gyro in the code here if you can select it as an alternative to PT1, right?  I'll pull it again and make sure, but I think so.
> 
>     * full-disclosure, there are no plans to release any further EmuFlight . i just thought this was interesting after noticing it was tested on @iNavFlight
I understand.  This version is improved beyond the current iNav one, and I intend to push the update to iNav as well.  Thank you for considering to merge!
> 
>     * it would require Configurator support as well, but it is currently impossible to build configurator as the OSX builds are broken.
I understand.  Thank you for noting this!!

pjpei commented:
ah wow, even my comments were incorrect, sorry. Thanks for fixing!

Quick-Flash commented:
Id like to see it as a CLI or OSD thing that can be added independently of the other gyro filters.

nerdCopter commented:
Closing this PR without merge. The LULU filter concept is preserved in issue #1276 for future reference.
